File: ios/pqp/Sources/Core/OnboardingAPI.swift
import Foundation

/// The server an invite opens, as a signed-out person may see it.
///
/// `GET /api/public/invites/:code` answers `{ invite: PublicInvitePreview }`:
/// a name, a picture and a head count, and nothing that names a member. It is
/// what lets the welcome screen say "You're invited to {server}" before there
/// is an account, and what the first-run "you" step shows as the room waiting.
struct PublicInvitePreview: Codable, Equatable, Sendable {
    let serverName: String
    let iconUrl: String?
    let memberCount: Int
}

/// What a Discord template turns into, trimmed to what a phone shows before
/// creating it. The server sends a much larger plan (`DiscordImportPlan` in
/// `packages/shared/src/discord-import.ts`); decoding only these fields keeps a
/// server-side addition from ever breaking the preview.
struct DiscordImportPreview: Decodable, Equatable, Sendable {
    struct Channel: Decodable, Equatable, Sendable {
        let name: String
        /// `"text" | "voice" | "category"`.
        let type: String
        let isPrivate: Bool
    }

    struct Role: Decodable, Equatable, Sendable {
        let name: String
    }

    let serverName: String
    let isDirty: Bool
    let channels: [Channel]
    let roles: [Role]

    var textCount: Int { channels.filter { $0.type == "text" }.count }
    var voiceCount: Int { channels.filter { $0.type == "voice" }.count }
    var categoryCount: Int { channels.filter { $0.type == "category" }.count }
}

/// `POST /api/import/discord/apply`: the server it made and the invite it
/// minted in the same transaction, so the ready step never has to ask twice.
struct DiscordImportResult: Decodable, Sendable {
    let server: Server
    let invite: Invite
}

extension APIClient {
    /// The invite preview, with NO Authorization header on purpose.
    ///
    /// The route is public and sits before the Bearer resolution on the server;
    /// a stale token must not be able to turn a preview into a 401 (pitfall 16
    /// in CLAUDE.md is exactly that shape). Every failure is "no preview", never
    /// an error: 404 (unknown, revoked, expired, or an API without the route
    /// yet), 429 (the anonymous limiter), a dropped network, or a body of the
    /// wrong shape. The screen then shows its generic copy and the flow still
    /// works, because the code itself is stashed in `PendingInvite`.
    nonisolated static func publicInvitePreview(
        code: String,
        backend: Backend = .current
    ) async -> PublicInvitePreview? {
        guard !code.isEmpty else { return nil }
        let url = backend.apiBaseURL
            .appendingPathComponent("api/public/invites")
            .appendingPathComponent(code)
        var request = URLRequest(url: url, timeoutInterval: 8)
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        guard let (data, response) = try? await URLSession.shared.data(for: request),
              let http = response as? HTTPURLResponse,
              (200..<300).contains(http.statusCode)
        else { return nil }
        return parsePublicInvitePreview(data)
    }

    /// `{ invite: { serverName, iconUrl, memberCount } }`, or nil. A blank name
    /// is no preview: "You're invited to " with nothing after it is worse than
    /// the generic welcome.
    nonisolated static func parsePublicInvitePreview(_ data: Data) -> PublicInvitePreview? {
        struct Envelope: Decodable { let invite: PublicInvitePreview }
        guard let preview = try? JSONDecoder().decode(Envelope.self, from: data).invite,
              !preview.serverName.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
        else { return nil }
        return preview
    }

    /// One field of `PATCH /api/me` at a time, so a rename can be told apart
    /// from a failed avatar. Absent keys are left alone server-side.
    func updateMe(displayName: String? = nil, username: String? = nil) async throws -> CurrentUser {
        struct Body: Encodable {
            let displayName: String?
            let username: String?
        }
        return try await patch("/api/me", body: Body(displayName: displayName, username: username))
    }

    /// A preset picture. The server accepts any http(s) URL here, which is how
    /// the web's presets are stored too, so a preset chosen on a phone reads
    /// the same in a browser.
    func setAvatarURL(_ url: String) async throws -> CurrentUser {
        struct Body: Encodable { let avatarUrl: String }
        return try await patch("/api/me", body: Body(avatarUrl: url))
    }

    /// An invite with an explicit lifetime. The ready step asks for a week
    /// with no use cap, the same invite the web wizard mints, so the note under
    /// the link ("Good for 7 days, unlimited uses") is true on both.
    func createInvite(serverId: String, expiresInHours: Int) async throws -> Invite {
        struct Body: Encodable { let expiresInHours: Int }
        struct Response: Decodable { let invite: Invite }
        let response: Response = try await post(
            "/api/servers/\(serverId)/invites", body: Body(expiresInHours: expiresInHours)
        )
        return response.invite
    }

    /// Close the first-run wizard for good, on every device. A one-key patch,
    /// for the same reason `dismissFirstRun` is one: the server merges the
    /// preferences blob one level deep, so naming a single key leaves the rest.
    func markOnboarded(at stamp: String) async throws -> UserPreferences {
        struct Body: Encodable { let onboardedAt: String }
        let response: PreferencesResponse = try await patch(
            "/api/me/preferences", body: Body(onboardedAt: stamp)
        )
        return response.preferences
    }

    /// The ids of every room this account is in, before it makes a new one.
    /// Nil when the list cannot be read, which switches recovery off rather
    /// than guessing.
    func serverIdsSnapshot() async -> Set<String>? {
        guard let servers = try? await servers() else { return nil }
        return Set(servers.map(\.id))
    }

    /// The room a create made, when the create's response was lost.
    ///
    /// The request may have reached the server and only the answer gone
    /// missing, and trying again would make a second room. So the caller takes
    /// a snapshot of its rooms BEFORE creating and, on a transport failure,
    /// asks for one it owns, with exactly that name, that was not there
    /// before. No clock is involved, so a phone whose time is off cannot make
    /// an old room look new, and an old room of the same name is excluded by
    /// construction.
    func serverCreatedSince(
        _ before: Set<String>,
        named name: String,
        ownerId: String
    ) async -> Server? {
        guard let servers = try? await servers() else { return nil }
        let fresh = servers.filter {
            !before.contains($0.id) && $0.ownerId == ownerId && $0.name == name
        }
        // Exactly one, or it is not ours to guess between.
        return fresh.count == 1 ? fresh[0] : nil
    }

    func previewDiscordImport(source: String) async throws -> DiscordImportPreview {
        struct Body: Encodable { let source: String }
        return try await post("/api/import/discord/preview", body: Body(source: source))
    }

    func applyDiscordImport(source: String) async throws -> DiscordImportResult {
        struct Body: Encodable { let source: String }
        return try await post("/api/import/discord/apply", body: Body(source: source))
    }
}

File: android/app/src/main/kotlin/gg/pqp/app/onboarding/OnboardingApi.kt
package gg.pqp.app.onboarding

import gg.pqp.app.core.ApiClient
import gg.pqp.app.core.Backend
import gg.pqp.app.core.Channel
import gg.pqp.app.core.Invite
import gg.pqp.app.core.Me
import gg.pqp.app.core.PqpJson
import gg.pqp.app.core.ServerSummary
import java.io.IOException
import kotlin.coroutines.resume
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.serialization.Serializable
import okhttp3.Call
import okhttp3.Callback
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response

/**
 * The endpoints first run needs and nothing else on Android called yet.
 *
 * Extensions on [ApiClient] rather than members, the way `AccountApi.kt` adds
 * its two: they are one feature's calls, and the client's body stays about
 * plumbing.
 */

/** `PATCH /api/me`. Only the keys that changed go on the wire. */
@Serializable
data class UpdateMeRequest(
    val displayName: String? = null,
    val username: String? = null,
    val avatarUrl: String? = null,
)

suspend fun ApiClient.updateMe(request: UpdateMeRequest): Me {
    val body = PqpJson.encodeToString(UpdateMeRequest.serializer(), request)
    return decode(
        execute(
            Request.Builder()
                .url(url("/api/me"))
                .patch(body.toRequestBody(ApiClient.JSON_MEDIA_TYPE)),
        ),
    )
}

@Serializable
private data class OnboardedPatch(val onboardedAt: String)

/**
 * `PATCH /api/me/preferences` with `onboardedAt`, which closes the wizard on
 * every device. The body is the merged preferences and nothing reads it.
 */
suspend fun ApiClient.markOnboarded(nowIso: String) {
    val body = PqpJson.encodeToString(OnboardedPatch.serializer(), OnboardedPatch(nowIso))
    execute(
        Request.Builder()
            .url(url("/api/me/preferences"))
            .patch(body.toRequestBody(ApiClient.JSON_MEDIA_TYPE)),
    ).close()
}

// ------------------------------------------------------------ invite preview

/** `publicInvitePreviewSchema` in `packages/shared/src/api.ts`. */
@Serializable
data class InvitePreview(
    val serverName: String,
    val iconUrl: String? = null,
    val memberCount: Int = 0,
)

@Serializable
private data class InvitePreviewResponse(val invite: InvitePreview? = null)

/**
 * `GET /api/public/invites/:code`: the server an invite opens, before the
 * person has an account.
 *
 * NO AUTH HEADER, and deliberately not through [ApiClient.execute], which
 * attaches one. The route is public and sits before the Bearer resolution, so
 * a stale token could only turn a preview into a 401 (pitfall 16 in
 * `CLAUDE.md` is that exact shape).
 *
 * Every failure is "no preview", never an error: 404 (unknown, revoked,
 * expired, or an API without the route yet), 429 from the anonymous limiter,
 * a dropped connection, or a body that is not the shape above. The screens
 * then show their generic copy, and nothing else depends on it.
 */
suspend fun fetchInvitePreview(
    http: OkHttpClient,
    code: String,
    apiUrl: String = Backend.apiUrl,
): InvitePreview? {
    if (!gg.pqp.app.invites.InviteLinks.isUsableCode(code)) return null
    val url = "$apiUrl/api/public/invites/$code".toHttpUrlOrNull() ?: return null
    val call = http.newCall(
        Request.Builder().url(url).header("Accept", "application/json").get().build(),
    )
    val raw = suspendCancellableCoroutine<String?> { continuation ->
        continuation.invokeOnCancellation { call.cancel() }
        call.enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                if (continuation.isActive) continuation.resume(null)
            }

            override fun onResponse(call: Call, response: Response) {
                val text = response.use {
                    if (it.isSuccessful) runCatching { it.body.string() }.getOrNull() else null
                }
                if (continuation.isActive) continuation.resume(text)
            }
        })
    } ?: return null
    return parseInvitePreview(raw)
}

/** `{ invite: InvitePreview }`, validated; anything else is no preview. */
fun parseInvitePreview(raw: String): InvitePreview? =
    runCatching { PqpJson.decodeFromString(InvitePreviewResponse.serializer(), raw).invite }
        .getOrNull()
        ?.takeIf { it.serverName.isNotBlank() }

// ------------------------------------------------------------ discord import

@Serializable
private data class DiscordImportSource(val source: String)

/**
 * The part of `DiscordImportPlan` (`packages/shared/src/discord-import.ts`)
 * the preview draws. The server keeps the rest and re-derives it on apply
 * from the same source, so nothing here is sent back.
 */
@Serializable
data class DiscordImportPlan(
    val serverName: String,
    val iconUrl: String? = null,
    val channels: List<DiscordImportChannel> = emptyList(),
    val roles: List<DiscordImportRole> = emptyList(),
)

@Serializable
data class DiscordImportChannel(
    val templateId: Long,
    val parentTemplateId: Long? = null,
    val type: String,
    val name: String,
    val position: Int = 0,
    val isPrivate: Boolean = false,
)

@Serializable
data class DiscordImportRole(val name: String)

@Serializable
data class DiscordImportResult(
    val server: ServerSummary,
    val channels: List<Channel> = emptyList(),
    val invite: Invite? = null,
)

suspend fun ApiClient.previewDiscordImport(source: String): DiscordImportPlan =
    postJson("/api/import/discord/preview", source)

suspend fun ApiClient.applyDiscordImport(source: String): DiscordImportResult =
    postJson("/api/import/discord/apply", source)

private suspend inline fun <reified T> ApiClient.postJson(path: String, source: String): T {
    val body = PqpJson.encodeToString(DiscordImportSource.serializer(), DiscordImportSource(source))
    return decode(
        execute(
            Request.Builder()
                .url(url(path))
                .post(body.toRequestBody(ApiClient.JSON_MEDIA_TYPE)),
        ),
    )
}
